pkg/mysqlsh/mysqlsh.go

```diff
@@ -0,0 +1,173 @@
+package mysqlsh
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"regexp"
+
+	"github.com/pkg/errors"
+	k8sexec "k8s.io/utils/exec"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/percona/percona-server-mysql-operator/pkg/innodbcluster"
+)
+
+type mysqlsh struct {
+	uri  string
+	exec k8sexec.Interface
+}
+
+var sensitiveRegexp = regexp.MustCompile(":.*@")
+
+const defaultIPAllowList = "127.0.0.1/8,::1/128"
+
+func New(e k8sexec.Interface, uri string) *mysqlsh {
+	return &mysqlsh{exec: e, uri: uri}
+}
+
+func (m *mysqlsh) run(ctx context.Context, cmd string) error {
+	var errb, outb bytes.Buffer
+
+	args := []string{"--uri", m.uri, "-e", cmd}
+
+	c := m.exec.CommandContext(ctx, "mysqlsh", args...)
+	c.SetStdout(&outb)
+	c.SetStderr(&errb)
+
+	if err := c.Run(); err != nil {
+		sout := sensitiveRegexp.ReplaceAllString(outb.String(), ":*****@")
+		serr := sensitiveRegexp.ReplaceAllString(errb.String(), ":*****@")
+		return errors.Wrapf(err, "run %s, stdout: %s, stderr: %s", cmd, sout, serr)
+	}
+
+	return nil
+}
+
+func (m *mysqlsh) ConfigureInstance(ctx context.Context, instance string) error {
+	cmd := fmt.Sprintf(
+		"dba.configureInstance('%s', {'interactive': false, 'clearReadOnly': true})",
+		instance,
+	)
+
+	if err := m.run(ctx, cmd); err != nil {
+		return errors.Wrap(err, "configure instance")
+	}
+
+	return nil
+}
+
+func (m *mysqlsh) AddInstance(ctx context.Context, clusterName, instance, podIp string) error {
+	opts := struct {
+		Interactive    bool   `json:"interactive"`
+		RecoveryMethod string `json:"recoveryMethod"`
+		WaitRecovery   int    `json:"waitRecovery"`
+		IpAllowList    string `json:"ipAllowList"`
+	}{
+		Interactive:    false,
+		RecoveryMethod: "clone",
```

Comment:
Can we use auto?
`auto: let Group Replication choose whether or not a full snapshot has to be taken, based on what the target server supports and the group_replication_clone_threshold sysvar. This is the default value. A prompt will be shown if not possible to automatically determine a convenient way forward. If the interaction is disabled, the operation will be cancelled instead.`

You can create separate task for research,

Reply:
https://jira.percona.com/browse/K8SPS-101